odoo/sql_db.py

```diff
@@ -358,32 +348,36 @@ def __init__(self, pool: ConnectionPool, dbname: str, dsn: dict):
         self._cnx: PsycoConnection = pool.borrow(dsn)
         self._obj: psycopg2.extensions.cursor = self._cnx.cursor()
         if _logger.isEnabledFor(logging.DEBUG):
-            self.__caller = frame_codeinfo(currentframe(), 2)
+            self.__caller: tuple[str, str | int] | None = frame_codeinfo(currentframe(), 2)
         else:
-            self.__caller = False
+            self.__caller = None
         self._closed = False   # real initialization value
-        # See the docstring of this class.
-        self.connection.set_isolation_level(ISOLATION_LEVEL_REPEATABLE_READ)
-        self.connection.set_session(readonly=pool.readonly)
+        self._cnx.set_session(
+            # See the docstring of this class.
+            isolation_level=ISOLATION_LEVEL_REPEATABLE_READ,
+            readonly=pool.readonly,
+            autocommit=False,
+        )
 
         if os.getenv('ODOO_FAKETIME_TEST_MODE') and self.dbname in tools.config['db_name']:
             self.execute("SET search_path = public, pg_catalog;")
             self.commit()  # ensure that the search_path remains after a rollback
 
-    def __build_dict(self, row: tuple) -> dict[str, typing.Any]:
+    def dictfetchone(self) -> dict[str, typing.Any] | None:
         description = self._obj.description
         assert description, "Query does not have results"
-        return {column.name: row[index] for index, column in enumerate(description)}
-
-    def dictfetchone(self) -> dict[str, typing.Any] | None:
         row = self._obj.fetchone()
-        return self.__build_dict(row) if row else None
+        return {column.name: row[index] for index, column in enumerate(description)} if row else None
 
     def dictfetchmany(self, size) -> list[dict[str, typing.Any]]:
-        return [self.__build_dict(row) for row in self._obj.fetchmany(size)]
+        description = self._obj.description
+        assert description, "Query does not have results"
+        return [{column.name: row[index] for index, column in enumerate(description)} for row in self._obj.fetchmany(size)]
 
     def dictfetchall(self) -> list[dict[str, typing.Any]]:
-        return [self.__build_dict(row) for row in self._obj.fetchall()]
+        description = self._obj.description
+        assert description, "Query does not have results"
+        return [{column.name: row[index] for index, column in enumerate(description)} for row in self._obj.fetchall()]
```

Comment:
nitpicking:
Intuitively, I would write it like this for performance but not sure it is worth it: 
```suggestion
        comumn_names = [column.name for column in description]
        return [dict(zip(comumn_names, row)) for row in self._obj.fetchmany(size)]
```